### PR TITLE
[train] Add documentation for using metadata argument to save preprocessors

### LINKED ISSUE #38614: [docs] Document metadata argument to Trainer
Document https://github.com/ray-project/ray/issues/38288

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This documents how to use the metadata arg for preserving preprocessor state (in addition to the pydoc). Add a util method to make this less ugly for preprocessor pickling.

## Related issue number

Closes https://github.com/ray-project/ray/issues/38614

### CODE REVIEW

File: doc/source/train/user-guides/data-loading-preprocessing.rst [hunk 1]
```diff
@@ -608,14 +618,6 @@ Transformations that you want run per-epoch, such as randomization, should go af
 
     ...
 
-Adding CPU-only nodes to your cluster
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (ericl) [inline doc/source/train/user-guides/data-loading-preprocessing.rst:612]: This is more advanced, so move to after prefetching section.

File: python/ray/data/preprocessor.py [hunk 2]
```diff
@@ -326,3 +327,14 @@ def preferred_batch_format(cls) -> BatchFormat:
         path is the most optimal.
         """
         return BatchFormat.PANDAS
+
+    @DeveloperAPI
+    def pickle(self) -> str:
+        """Return this preprocessor pickled as a string."""
+        return pickle.dumps(self, 0).decode("ascii")
```
Comment (pcmoritz) [inline python/ray/data/preprocessor.py:334]: I would leave this as the default protocol (i.e. pickle.DEFAULT_PROTOCOL) since protocol 0 is really space inefficient (some kind of text protocol I believe). I don't think there is any advantage of using 0 here but lmk if you think there is :)

**Review by pcmoritz — APPROVED** [review]
Nice!

File: python/ray/data/preprocessor.py (hunk 2 above)
Comment (ericl) [reply python/ray/data/preprocessor.py:334]: I believe we have to use this protocol if we want to ensure it's ASCII decodable. Alternatively we could base64 the string, though this way seems a bit less opaque.

Comment (pcmoritz) [reply python/ray/data/preprocessor.py:334]: Ah lol, but it also doesn't work --
```python
pickle.dumps('á', 0).decode("ascii")
```
gives an error (this is basically a Python bug https://bugs.python.org/issue2980). Why don't we use the default protocol and base64, then we are safe (and not too verbose) :)

File: doc/source/train/user-guides/data-loading-preprocessing.rst [hunk 3]
```diff
@@ -532,29 +535,46 @@ You can use this with Ray Train Trainers by applying them on the dataset before
...
+    metadata = my_trainer.fit().checkpoint.get_metadata()
+    print(Preprocessor.unpickle(metadata["preprocessor_pkl"]))
 
 
 .. testoutput::
     :hide:
 
     ...
 
+In this example, we persist the fitted preprocessor using the ``Trainer(metadata={...})`` constructor argument. This arg specifies a dict that will available as ``checkpoint.get_metadata()`` for checkpoints saved from the Trainer. This enables recreation of the fitted preprocessor for use for inference.
```
Comment (matthewdeng) [inline doc/source/train/user-guides/data-loading-preprocessing.rst:577]: Also mention here the `train.get_context().get_metadata()` part?

File: python/ray/data/preprocessor.py (hunk 2 above)
Comment (matthewdeng) [reply python/ray/data/preprocessor.py:334]: Ah one thing I learned from @justinvyu is to use these utils:

```
from ray._private.utils import binary_to_hex, hex_to_binary
```

https://github.com/ray-project/ray/blob/41e34a15cea055f3f4bbacae0b7bc46dd5b211b4/python/ray/train/_internal/framework_checkpoint.py#L28-L45

Would that work?

**Review by matthewdeng — APPROVED** [review]
Nice!

File: python/ray/data/preprocessor.py (hunk 2 above)
Comment (ericl) [reply python/ray/data/preprocessor.py]: Fixed.

Comment (ericl) [reply python/ray/data/preprocessor.py]: @matthewdeng that would work, but seems the `base64` module also works.

**Comment by pcmoritz** [comment]
If we call this serialize / deserialize we should point out in the docs that this uses pickle under the hood and the format is therefore not stable. Otherwise it will catch people by surprise.

**Review by pcmoritz — APPROVED** [review]

**Comment by ericl** [comment]
Hmm, seems I can't merge this since it depends on the new storage flag being enabled. @justinvyu should I wait for that to happen?

**Comment by justinvyu** [comment]
Yeah, all the doc tests are lumped together, so this may be easier to do in one go while transitioning all doc examples.

**Comment by matthewdeng** [comment]
@ericl This should now be unblocked - the doctest jobs are now running with the new persistence flag enabled (https://github.com/ray-project/ray/pull/38876).

**Review by richardliaw — APPROVED** [review]

File: doc/source/train/user-guides/data-loading-preprocessing.rst [hunk 4]
```diff
@@ -527,10 +527,12 @@ You can use this with Ray Train Trainers by applying them on the dataset before
 .. testcode::
 
     import numpy as np
+    from tempfile import TemporaryDirectory
 
     import ray
     from ray import train
     from ray.train import ScalingConfig
+    from ray.train._checkpoint import Checkpoint
```
Comment (richardliaw) [inline doc/source/train/user-guides/data-loading-preprocessing.rst:535]: interesting

**Comment by ericl** [comment]
Only failing test seems to be an unrelated lightning trainer test.

File: doc/source/train/user-guides/data-loading-preprocessing.rst (hunk 4 above)
Comment (matthewdeng) [inline doc/source/train/user-guides/data-loading-preprocessing.rst:535]:
```suggestion
    from ray.train import Checkpoint
```

**Review by scv119 — APPROVED** [review]
